pkg/asset/manifests/machine-api-operator.go

```diff
@@ -32,36 +28,8 @@ var _ asset.WritableAsset = (*machineAPIOperator)(nil)
 // TODO(enxebre): move up to github.com/coreos/tectonic-config (to install-config? /rchopra)
 type maoOperatorConfig struct {
 	metav1.TypeMeta `json:",inline"`
-	TargetNamespace string           `json:"targetNamespace"`
-	APIServiceCA    string           `json:"apiServiceCA"`
-	Provider        string           `json:"provider"`
-	AWS             *awsConfig       `json:"aws"`
-	Libvirt         *libvirtConfig   `json:"libvirt"`
-	OpenStack       *openstackConfig `json:"openstack"`
-}
-
-type libvirtConfig struct {
-	ClusterName string `json:"clusterName"`
-	URI         string `json:"uri"`
-	NetworkName string `json:"networkName"`
-	IPRange     string `json:"iprange"`
-	Replicas    int    `json:"replicas"`
-}
-
-type awsConfig struct {
-	ClusterName      string `json:"clusterName"`
-	ClusterID        string `json:"clusterID"`
-	Region           string `json:"region"`
-	AvailabilityZone string `json:"availabilityZone"`
-	Image            string `json:"image"`
-	Replicas         int    `json:"replicas"`
-}
-
-type openstackConfig struct {
-	ClusterName string `json:"clusterName"`
-	ClusterID   string `json:"clusterID"`
-	Region      string `json:"region"`
-	Replicas    int    `json:"replicas"`
+	TargetNamespace string `json:"targetNamespace"`
+	Provider        string `json:"provider"`
```

Comment:
Do we need these two?  Can MAO install them on its own?  It already has namespace opinions [here][1].  And the provider information is already included in the worker machine-set:

```console
$ openshift-install --dir testing create manifests
$ cat testing/tectonic/99_openshift-cluster-api_worker-machineset.yaml

apiVersion: cluster.k8s.io/v1alpha1
kind: MachineSet
metadata:
  ...
spec:
  replicas: 1
  selector:
    ...
  template:
    metadata:
      ...
    spec:
      providerConfig:
        value:
          apiVersion: libvirtproviderconfig.k8s.io/v1alpha1
          kind: LibvirtMachineProviderConfig
          ...
      versions:
        ...
```

And also master machine entry:

```console
$ cat testing/tectonic/99_openshift-cluster-api_master-machines.yaml
kind: List
apiVersion: v1
metadata:
  resourceVersion: ""
  selfLink: ""
items:
- apiVersion: cluster.k8s.io/v1alpha1
  kind: Machine
  metadata:
    ...
  spec:
    providerConfig:
      value:
        apiVersion: libvirtproviderconfig.k8s.io/v1alpha1
        kind: LibvirtMachineProviderConfig
        ...
    versions:
      ...
```

[1]: https://github.com/openshift/machine-api-operator/blob/071531d180e0b95ec8cea8136264f5eb5695f268/install/0000_50_machine-api-operator_00_namespace.yaml#L4

Reply:
> ... the provider is needed so the operator knows which provider controller to run....

Don't both `.spec.providerConfig.value.apiVersion` and `.spec.providerConfig.value.kind` contain the provider information you need (see my examples above, where they are `libvirtproviderconfig.k8s.io/v1alpha1` and `LibvirtMachineProviderConfig` respectively)?

Reply:
I'm not sure how you mean here . The operator needs to know the provider so it runs a different controller depending on it. Then that controller understand only its specific kind. There's no global controller which decode the kind and choose and implementation based on it, currently the cluster api requires to choose a provider controller ahead of time. Let me know if does not make sense

Reply:
> The operator needs to know the provider so it runs a different provider controller depending on it. Then that controller watches machines and understand only its specific kind. There's no global controller which decode the kind and choose an implementation based on it at runtime; currently the cluster api requires to choose a provider controller ahead of time.

Ah, maybe we should grow this ;).  As it stands, can't we do the Go equivalent of:

```sh
oc get -o jsonpath --template '{.items[0].spec.template.spec.providerConfig.value.kind}' -n openshift-cluster-api machinesets
```

from the operator to figure out which controller to launch?

Reply:
@wking do you want the operator to deploy controllers based on the machinesets that were pushed to api? :P I don't think that's a good idea.

@enxebre almost all operators are using the `kube-system/cluster-config-v1` configmap'ss key `install-config` to discovery configuration like the platform. That can be a follow up to remove any mao specific code from installer as it is not required during bootstrap.

Reply:
> @wking do you want the operator to deploy controllers based on the machinesets that were pushed to api? :P I don't think that's a good idea.

Yeah that's what I want.  But I'm not very familiar with this space, maybe it's crazy.  Is there a big difference between the operator choosing based on a `machineAPIOperatorConfig` and the operator choosing based on the worker `MachineSet`?  They seem similar to me, except the latter would be more DRY.  And it also sets us up for "this is mostly an AWS cluster, but toss some workers up on GCP" at some point in the distant future.

> @enxebre almost all operators are using the kube-system/cluster-config-v1 configmap's key install-config to discover configuration like the platform.

But weren't we planning on dropping that?  I'll see if I can find a reference.

Reply:
> And it also sets us up for "this is mostly an AWS cluster, but toss some workers up on GCP"

I think we are getting too ahead of ourselves :P. Our machines only talk on internal load balancers, this will not work. also none of the other components are multi cloud aware and i don't think they should be... due to simpler decision boundaries..

> But weren't we planning on dropping that? I'll see if I can find a reference.

We might drop it in favor of something else but there will always be some discovery available for operator.

Reply:
hey @wking @abhinavdahiya There's too many moving pieces yet (downstream and upstream) that makes me hesitant on thinking about enabling a discovery which requires to run at least on object before the controller which watches it. I can't see any effective immediate benefit whereas it would get as to a path with loads of open questions and a bunch of new cases that we'll have to contemplate and which are difficult to reason about at this point (what happen if the operator see two objects for different clouds? use the first one? do they map to the same cluster object? or do we support simultaneous clouds? if so that's a whole different story, etc, etc... ). The cluster api itself might support a more dynamic system for running the controller eventually or we could think of it when the times come but it's just too soon. I think for now using https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L34 make sense. I'll update the operator to consume the installconfig so totally disconnect it from this configmap, then I'll update this PR to drop it all

Reply:
>  I'll update the operator to consume the installconfig...

I dunno how future-proof this will be, but I'm completely fine with it as a short-term solution.  And who knows, maybe it will last longer than I think ;).

Reply:
> > @enxebre almost all operators are using the kube-system/cluster-config-v1 configmap's key install-config to discover configuration like the platform.
>
> But weren't we planning on dropping that? I'll see if I can find a reference.

[Here][1]'s the reference I was thinking of.

[1]: https://github.com/openshift/installer/pull/465#issuecomment-429978860